File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-0.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left = 0
        max_len = 0
        mapp = {}

        for right in range(len(s)):
            ch = s[right]

            if ch in mapp and mapp[ch] >= left:
                left = mapp[ch] + 1

            mapp[ch] = right
            max_len = max(max_len, right - left + 1)

        return max_len

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<int, int> m = {};
        int left = 0;
        int max_freq = 0;
        int max_len = 0;

        for(int right = 0; right < s.size(); right++){
            m[s[right]]++;
            max_freq = max(max_freq, m[s[right]]);

            int r = (right - left + 1) - max_freq;

            if(r > k){
                m[s[left]]--;
                left++;
            }
            
            max_len = max(max_len, right - left + 1);
        }

        return max_len;
    }
};
